src/pubd/repository.rs

```diff
@@ -151,22 +167,50 @@ impl RepositoryContentProxy {
         self.write(|content| content.session_reset(config))
     }
 
+
     fn write<F: FnOnce(&mut RepositoryContent) -> KrillResult<()>>(&self, op: F) -> KrillResult<()> {
+        // If there is any existing content, then we can assume that the cache
+        // has it - because it's initialized when we read the content during
+        // initialization.
         let store = self.store.write().unwrap();
-        let mut content: RepositoryContent = store.get(&self.key)?.ok_or(Error::RepositoryServerNotInitialized)?;
+        let mut cache = self.cache.write().unwrap();
+
+        let content_arc: &mut Arc<RepositoryContent> = cache.as_mut().ok_or(Error::RepositoryServerNotInitialized)?;
+        let content = Arc::<RepositoryContent>::get_mut(content_arc).unwrap();
```

Comment:
I think this `unwrap` will panic under certain conditions. `Arc::get_mut` only succeeds if there is exactly one clone out there. Yet you clone the arc in `read_content_cache`. So if any of the methods calling `write` are called while the clone returned by `read_content_cache` is still alive, you’ll get a panic.

I think you don’t need the arc here at all? It looks like all the methods that call `read_content` only call a method on the content anyway, so this could probably just return the `RwLockReadGuard` returned by the lock.

Reply:
Good point. I removed the `Arc`. And then I changed things around a bit and added `read` instead of `read_content` which takes a closure which it can execute within a read 'transaction' without the need to clone.